src/component/Slider/Slider.jsx

```diff
@@ -0,0 +1,275 @@
+    if ((infiniteLoop || currentIndex < length - slidesPerView) && currentIndex !== (length + slidesPerView)) {
+      setCurrentIndex((index) => (index + 1));
+      setCurrentIndicator((index) => (index + 1));
+    }
+  };
+
+  const onClickIndicator = e => {
+    const value = parseInt(e.target.value, 10);
+    setCurrentIndex(value);
+    setCurrentIndicator(value);
+  };
+
+  const onTouchStart = e => {
+    let clientDownX = e.touches[0].clientX;
+
+    setScrollingState({
+      ...scrollingState,
+      isTransition: false,
+      isScrolling: true,
+      clientX: clientDownX,
+    })
+  }
+
+  const onMouseDown = e => {
+    e.preventDefault();
+    let clientDownX = e.clientX;
+
+    setScrollingState({
+      ...scrollingState,
+      isTransition: false,
+      isScrolling: true,
+      clientX: clientDownX,
+    })
+  }
+
+  const handleSlideMove = (currenClientX) => {
+    const { isScrolling, clientX } = scrollingState;
+
+    if (isScrolling) {
+      const scrX = ((clientX - currenClientX )/carouselRef.current.clientWidth) ;
+      setScrollingState({
+        ...scrollingState,
+        scrollX: scrX
+      });
+    }
+  }
+
+  const onTouchMove = e => {
+    let currenClientX = e.touches[0].clientX;
+    handleSlideMove(currenClientX);
+    }
+
+  const onMouseMove = e => {
+    e.preventDefault();
+    let currenClientX = e.clientX;
+    handleSlideMove(currenClientX);
+  }
+
+  const changeSlideHandle = () => {
+    if (scrollingState.scrollX > 0.3) {
+      if (currentIndex === (length - slidesPerView) && !infiniteLoop) {
+        currentSlideShow();
+        setScrollingState({
+          ...scrollingState,
+          isTransition: true,
+          isScrolling: false,
+          scrollX: 0,
+          clientX: 0,
+        })
+        return;
+      }
+
+      onClickNext();
+    } else if (scrollingState.scrollX < -0.3) {
+      onClickPrev();
+    } else {
+      currentSlideShow();
+    }
+    setScrollingState({
+      ...scrollingState,
+      isTransition: true,
+      isScrolling: false,
+      scrollX: 0,
+      clientX: 0,
+    })
+  }
+
+  const currentSlideShow = () => {
+    setScrollingState({
+      ...scrollingState,
+      isTransition: true,
+      isScrolling: false,
+      scrollX: 0,
+    })
+  }
+
+  const onMouseUp = () => {
+    if (!scrollingState.isScrolling) {
+      return;
+    }
+    changeSlideHandle();
+  }
+
+  const onTouchEnd = () => {
+    if (!scrollingState.isScrolling) {
+      return;
+    }
+
+    changeSlideHandle();
+  }
+
+  const onMouseLeave = e => {
+    if (!scrollingState.isScrolling) {
+      return;
+    }
+
+    e.preventDefault();
+    changeSlideHandle();
+  }
+
+  const onTouchCancel = () => {
+    if (!scrollingState.isScrolling) {
+      return;
+    }
+
+    changeSlideHandle();
+  }
+
+  const renderPrevSlide = () => {
+    const output = [];
+    for (let i = 0; i < slidesPerView; i += 1) {
+      output.push(children[length - 1 - i]);
```

Comment:
Possible OutOfRange exception. Why are you so sure that calculated index will be in a range of children values

Reply:
You're right, I didn't think about of all the possibilities. So I corrected the condition for the loop.